src/home/homed-home.c

```diff
@@ -1210,6 +1225,29 @@ static int home_start_work(Home *h, const char *verb, UserRecord *hr, UserRecord
                         return r;
         }
 
+        if (blobs) {
+                blob_fd_arr = newa(int, hashmap_size(blobs));
+
+                /* homework needs to be able to tell the difference between blobs being null
+                 * (the fdmap field is completely missing) and it being empty (the field is an
+                 * empty object) */
+                r = json_variant_new_object(&fdmap, NULL, 0);
+                if (r < 0)
+                        return r;
+
+                HASHMAP_FOREACH_KEY(fd_ptr, blob_filename, blobs) {
+                        r = json_variant_set_field_integer(&fdmap, blob_filename, PTR_TO_FD(fd_ptr));
```

Comment:
hmpf, fds passed as json numbers is weird.

what about this: pass paths here, but make those paths `/proc/$PID/fd/$FD` instead, referring to homed's pid. that way we pin by fd but can communicate paths, and homework would just open() them

Reply:
(or maybe  not, just an idea)

Reply:
This opens a can of worms w.r.t. ownership of the FDs. Right now, the hashmap closes all the FDs when it goes out of scope. The map is allocated all the way up in the dbus code, and passed down the call stack to ultimately end up in home_start_work. Pretty much immediately after we fork off homework, we return all the way up the call stack back into the main loop, freeing the hashmap and thus closing all the FDs in the process. So the solution you mention is racy - the FD may no longer be around by the time homework is running and trying to open it

I suppose we can somehow signal from homework to homed that it can free the map, but that's getting really complicated. I suppose we can also transfer ownership of the blobs hashmap into the ongoing operation so that it gets free'd after homework is done, but I see little reason to keep that memory allocated in homed once we've passed it off into homework - it'll never get used again in homed. Any way you spin it, we'd need to keep the FDs open in homed at least until homework opens its own copies: passing the FDs across the `fork` is probably the best way to do it (since homework gets its own open copies immediately on startup)